### PR TITLE
Migrate league/tactician-bundle to symfony/messenger

### LINKED ISSUE #32786: Migration of tactician towards Symfony Messenger
Currently we're using a library called [league/tactician-bundle](https://packagist.org/packages/league/tactician-bundle) which provide a message bus.
Symfony (since the version 4) provides it's own component with a message bus which is [symfony/messenger](https://symfony.com/doc/current/messenger.html).
We should migrate to symfony/messenger to take advantage of the symfony ecosystem. Also it's more maintainable since this library evolves the same rythme than the other components. For example all the symfony components use the same php version for a given symfony version.

### PR DESCRIPTION
**Pull request by lartist**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Migrate league/tactician-bundle to symfony/messenger
| Type?             | improvement
| Category?         | CO
| BC breaks?        | yes
| Deprecations?     | no
| How to test?      | CI 🟢  and auto tests 🟢 
| Fixed ticket?     | Fixes #32786
| Related PRs       | 
| Sponsor company   |

### CODE REVIEW

File: src/Core/CommandBus/Parser/CommandDefinition.php [hunk 1]
```diff
@@ -1,84 +0,0 @@
...
- * @copyright Since 2007 PrestaShop SA and Contributors
- * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
- */
-
-namespace PrestaShop\PrestaShop\Core\CommandBus\Parser;
-
-/**
- * Transfers commands and queries definition data
- */
-class CommandDefinition
```
Comment (jolelievre) [inline src/Core/CommandBus/Parser/CommandDefinition.php]: Even if we use Messenger instead of tacticien this DTO is still relevant, we shouldn't remove this that is used by our command to list infos about CQRS command Instead this should remain but the command will probably need to be modified to be compatible with the new Messenger implementation

And the tricky part will be how to tell apart messenger handlers related to CQRS commands from other ones that may be used for other use cases in the future or by other frameworks

Comment (jolelievre) [reply src/Core/CommandBus/Parser/CommandDefinition.php]: I'm thinking of something to differentiate CQRS commands handlers:

Capture d’écran 2023-06-08 à 12 45 32

Maybe some other option could be added with the tag? Or we could introduce a custom tag `messenger.cqrs_handler` that would act the same as `messenger.message_handler` but at least we'd know it's part of the CQRS scope

**Review by jolelievre — CHANGES_REQUESTED** [review]
Migrating to messenger should not have impact on our Symfony commands used to list the CQRS command

The migration is only acceptable if we don't lose some features and we don't use the commands in a different (thankfully we already had an interface to abstract the implementation), except for service definitions it should be transparent/invisible

**Comment by M0rgan01** [comment]
- Withdrawal of inheritance between different commands (If my command 1 inherited from my command 2, then both handler 1 and handler 2 were called during the dispatch of command 1).

- There are breaking changes for modules that will remain on Tactician.

- The Symfony Messenger tags that enable proper command management are built within the "CommandAndQueryCollectorPass" class.

- Addition of 2 new attributes "AsQueryHandler" and "AsCommandHandler" used to define the handlers. We will no longer use tags, we will add an "autoconfigure" option to register handlers with the new attributes

**Comment by M0rgan01** [comment]
Tests: https://github.com/M0rgan01/ga.tests.ui.pr/actions/runs/5343755783

File: src/PrestaShopBundle/DependencyInjection/Compiler/CommandAndQueryCollectorPass.php [hunk 2]
```diff
@@ -67,4 +67,16 @@ private function findCommandsAndQueries(ContainerBuilder $container)
 
         return $commands;
     }
+
+    /**
+     * update messenger tags allowing the recognition of handlers by symfony
+     */
+    private function updateMessengerTags(ContainerBuilder $container, array $handlers): void
```
Comment (FabienPapet) [inline src/PrestaShopBundle/DependencyInjection/Compiler/CommandAndQueryCollectorPass.php]: We should add the tag in every service definition instead of creating this. That avoids complexity and maintenance.

**Review by FabienPapet — DISMISSED** [review]

File: src/PrestaShopBundle/DependencyInjection/Compiler/CommandAndQueryCollectorPass.php (hunk 2 above)
Comment (M0rgan01) [reply src/PrestaShopBundle/DependencyInjection/Compiler/CommandAndQueryCollectorPass.php]: I do not think so. If tomorrow symfony changes its tags, we will have to change the tags in one place. Also if we want to use symfony messenger for another use, there will be misunderstanding. Also it makes us gain in readability, we use 2 lines instead of 3, because the method is always the same

File: src/Core/CommandBus/MessengerCommandBusAdapter.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    public function __construct(private readonly MessageBusInterface $commandBus)
+    {
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handle($command)
+    {
+        try {
```
Comment (jolelievre) [inline src/Core/CommandBus/MessengerCommandBusAdapter.php]: You should use `Symfony\Component\Messenger\HandleTrait` to avoid duplicating its code, now I know the two methods have the same name but @FabienPapet just taught me that you can rename the methods from traits.

You'll find an example here https://andy-carter.com/blog/overriding-extending-a-php-trait-method

```php
class Example
{
	use FooTrait {
		bar as traitBar;
	}

	public function bar()
	{
		return $this->traitBar() . 'B';
	}
}
```

Comment (M0rgan01) [reply src/Core/CommandBus/MessengerCommandBusAdapter.php]: Thanks for the tips :+1:

**Comment by M0rgan01** [comment]
Tests: https://github.com/M0rgan01/ga.tests.ui.pr/actions/runs/5379225047

File: src/Core/CommandBus/MessengerCommandBusAdapter.php (hunk 3 above)
Comment (lartist) [reply src/Core/CommandBus/MessengerCommandBusAdapter.php]: @jo why not using the injected MessageBusInterface, it's currently useless !

File: src/Core/CommandBus/Middleware/CommandRegisterMiddleware.php [hunk 4]
```diff
@@ -26,56 +26,32 @@
...
 namespace PrestaShop\PrestaShop\Core\CommandBus\Middleware;
 
-use League\Tactician\Handler\CommandNameExtractor\CommandNameExtractor;
-use League\Tactician\Handler\Locator\HandlerLocator;
-use League\Tactician\Middleware;
 use PrestaShop\PrestaShop\Core\CommandBus\ExecutedCommandRegistry;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Handler\HandlersLocatorInterface;
+use Symfony\Component\Messenger\Middleware\MiddlewareInterface;
+use Symfony\Component\Messenger\Middleware\StackInterface;
```
Comment (lartist) [inline src/Core/CommandBus/Middleware/CommandRegisterMiddleware.php]: Here we are using Symfony in the Core namespace !

File: src/Core/CommandBus/MessengerCommandBusAdapter.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+namespace PrestaShop\PrestaShop\Core\CommandBus;
+
+use Symfony\Component\Messenger\Exception\HandlerFailedException;
+use Symfony\Component\Messenger\HandleTrait;
+use Symfony\Component\Messenger\MessageBusInterface;
+
+/**
+ * Class MessengerCommandBusAdapter is the Symfony Messenger CommandsBus implementation for PrestaShop's contract.
+ */
+final class MessengerCommandBusAdapter implements CommandBusInterface
```
Comment (lartist) [inline src/Core/CommandBus/MessengerCommandBusAdapter.php]: This classe should be in the Adapter namespace

File: src/Core/CommandBus/ExecutedCommandRegistry.php [hunk 6]
```diff
@@ -115,7 +115,7 @@ private function getTrace(): array
         $trace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, self::BACKTRACE_LIMIT);
 
         foreach ($trace as $step) {
-            if ($step['class'] === TacticianCommandBusAdapter::class
+            if ($step['class'] === MessengerCommandBusAdapter::class
```
Comment (lartist) [inline src/Core/CommandBus/ExecutedCommandRegistry.php]: It's not due to your PR as the problem was here before but I just want to mention that the MessengerCommandBusAdapter should be under Adapter namespace and then not using here (in the Core namespace)

**Review by boherm — COMMENTED** [review]
Nothing to add to what others have already said about your work.
Thank you very much for this PR!
I will approve when all conversations are resolved.

File: src/Core/CommandBus/MessengerCommandBusAdapter.php (hunk 3 above)
Comment (M0rgan01) [reply src/Core/CommandBus/MessengerCommandBusAdapter.php]: The injected MessageBusInterface is used in the HandleTrait

Comment (M0rgan01) [reply src/Core/CommandBus/MessengerCommandBusAdapter.php]: I will post a comment to clarify this.

Comment (lartist) [reply src/Core/CommandBus/MessengerCommandBusAdapter.php]: Ok but why are we using the trait instead of the injected MessageBusInterface only?
`$this->process($command)` instead of `$this->messageBus->handle($command)`

Comment (M0rgan01) [reply src/Core/CommandBus/MessengerCommandBusAdapter.php]: Using "$this->messageBus->dispatch($command)" provides a Envelope object, not the expected result directly.
The trait directly retrieves the result and ensures that there are no multiple handlers that handled the result, in addition to other verifications. See https://symfony.com/doc/current/messenger/handler_results.html

**Comment by M0rgan01** [comment]
Tests: https://github.com/M0rgan01/ga.tests.ui.pr/actions/runs/5473339359

**Review by boherm — DISMISSED** [review]

File: phpstan-baseline.neon [hunk 7]
```diff
@@ -1769,3 +1769,14 @@ parameters:
...
+		-
+			message: "#^Parameter \\#2 \\$configurator of method Symfony\\\\Component\\\\DependencyInjection\\\\ContainerBuilder::registerAttributeForAutoconfiguration\\(\\) expects callable#"
+			count: 2
+			path: src/PrestaShopBundle/DependencyInjection/Compiler/CommandAndQueryRegisterPass.php
+
+		-
+			message: "#^Call to an undefined method ReflectionType::getName\\(\\)#"
+			count: 1
+			path: src/PrestaShopBundle/DependencyInjection/Compiler/CommandAndQueryRegisterPass.php
+
```
Comment (jolelievre) [inline phpstan-baseline.neon]: I don't understand why we need these exceptions 🤔 I'm not sure but I guess fixing some types or adding some `instanceof` checks or PHPDoc should help fix those I can look into it if you want

File: src/Core/CommandBus/ExecutedCommandRegistry.php (hunk 6 above)
Comment (jolelievre) [reply src/Core/CommandBus/ExecutedCommandRegistry.php]: I agree, I'm think maybe this could be replaced with:

```suggestion
            if ($step['class'] instanceof CommandBusInterface::class
```

Comment (jolelievre) [reply src/Core/CommandBus/ExecutedCommandRegistry.php]: And the new `MessengerCommandBusAdapter` could indeed be moved into the Adapter namespace, or actually maybe more in PrestaShopBundle since it's a bridge between the core interface and the Symfony messenger component

File: src/Core/CommandBus/MessengerCommandBusAdapter.php (hunk 3 above)
Comment (jolelievre) [reply src/Core/CommandBus/MessengerCommandBusAdapter.php]: @lartist yeah I was surprized at first too, but when you look at the code of messenger it is not adapted to our use case which is to return the object the handler actually returned Although the messenger component already has implemented a trait that does exactly what we need so it makes sense to rely on it (you can look into the trait it should be clearer that my explanation)

The only technical issue was that both methods had the same names, but then @FabienPapet explained how we can actually rename the methods from a trait, which is a cool feature I had no idea existed!

File: src/Core/CommandBus/Middleware/CommandRegisterMiddleware.php (hunk 4 above)
Comment (jolelievre) [reply src/Core/CommandBus/Middleware/CommandRegisterMiddleware.php]: I also think this one could be placed into the PrestaShopBundle for two reasons:
- this is a middleware for the messenger component which is related to the symfony framework
- the purpose here is to register the executed commands which are then use in the Symfony debug toolbar

So both the technical and functional point of view tend to Symfony relations, hence it makes sense to move this into the bundle

File: src/Core/Domain/MailTemplate/CommandHandler/GenerateThemeMailTemplatesHandlerInterface.php [hunk 8]
```diff
@@ -29,9 +29,9 @@
 use PrestaShop\PrestaShop\Core\Domain\MailTemplate\Command\GenerateThemeMailTemplatesCommand;
 
 /**
- * Interface GenerateThemeMailsHandlerInterface
+ * Interface GenerateThemeMailTemplatesHandlerInterface
  */
-interface GenerateThemeMailTemplatesCommandHandlerInterface
```
Comment (jolelievre) [inline src/Core/Domain/MailTemplate/CommandHandler/GenerateThemeMailTemplatesHandlerInterface.php:34]: Why did you need to rename this interface? 🤔

File: src/Core/CommandBus/MessengerCommandBusAdapter.php (hunk 3 above)
Comment (lartist) [reply src/Core/CommandBus/MessengerCommandBusAdapter.php]: Does that mean we can't use middlewares with this ?

File: tests/Integration/Adapter/Category/QueryHandler/EditRootCategoryHandlerTest.php [hunk 9]
```diff
@@ -37,7 +37,7 @@
 class EditRootCategoryHandlerTest extends KernelTestCase
 {
     /**
-     * @var object|TacticianCommandBusAdapter|null
+     * @var object|MessengerCommandBusAdapter|null
```
Comment (jolelievre) [inline tests/Integration/Adapter/Category/QueryHandler/EditRootCategoryHandlerTest.php]:
```suggestion
     * @var object|CommandBusInterface|null
```

File: tests/Integration/Adapter/Category/QueryHandler/GetCategoryForEditingHandlerTest.php [hunk 10]
```diff
@@ -38,7 +38,7 @@
 class GetCategoryForEditingHandlerTest extends KernelTestCase
 {
     /**
-     * @var object|TacticianCommandBusAdapter|null
+     * @var object|MessengerCommandBusAdapter|null
```
Comment (jolelievre) [inline tests/Integration/Adapter/Category/QueryHandler/GetCategoryForEditingHandlerTest.php]:
```suggestion
     * @var object|CommandBusInterface|null
```

File: tests/Unit/Core/CommandBus/MessengerCommandBusAdapterTest.php [hunk 11]
```diff
@@ -26,16 +26,16 @@
...
+use PrestaShop\PrestaShop\Core\CommandBus\MessengerCommandBusAdapter;
+use Symfony\Component\Messenger\MessageBusInterface;
 
-class TacticianCommandBusAdapterTest extends TestCase
+class MessengerCommandBusAdapterTest extends TestCase
 {
     public function testIsValidImplementation()
     {
-        $commandBudAdapter = new TacticianCommandBusAdapter($this->createMock(CommandBus::class));
+        $commandBudAdapter = new MessengerCommandBusAdapter($this->createMock(MessageBusInterface::class));
```
Comment (jolelievre) [inline tests/Unit/Core/CommandBus/MessengerCommandBusAdapterTest.php]: Quite a useless test IMO but why not 😅

File: src/Core/Domain/MailTemplate/CommandHandler/GenerateThemeMailTemplatesHandlerInterface.php (hunk 8 above)
Comment (M0rgan01) [reply src/Core/Domain/MailTemplate/CommandHandler/GenerateThemeMailTemplatesHandlerInterface.php:34]: To follow the naming convention of other handler interfaces, they all end with *Handler Interface, rather than *CommandHandler Interface.

File: src/Core/CommandBus/ExecutedCommandRegistry.php [hunk 12]
```diff
@@ -114,8 +115,8 @@ private function getTrace(): array
     {
         $trace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, self::BACKTRACE_LIMIT);
 
-        foreach ($trace as $step) {
-            if ($step['class'] === TacticianCommandBusAdapter::class
+        foreach ($trace as $key => $step) {
+            if (is_a($step['class'], MessengerCommandBus::class, true)
                 && $step['function'] === 'handle'
```
Comment (jolelievre) [inline src/Core/CommandBus/ExecutedCommandRegistry.php]: Could you use the interface instead of the concrete class here please? Also could you reverse both conditions, I think it will be a bit more optimized to check for the function name first as `is_a` may require more resource than a simple string equality

```suggestion
            if ($step['function'] === 'handle'
                && is_a($step['class'], CommandBusInterface::class, true)
```

Comment (M0rgan01) [reply src/Core/CommandBus/ExecutedCommandRegistry.php]: Well seen, it will indeed be more optimized

File: src/Core/CommandBus/ExecutedCommandRegistry.php [hunk 13]
```diff
@@ -29,6 +29,7 @@
 namespace PrestaShop\PrestaShop\Core\CommandBus;
 
 use PrestaShop\PrestaShop\Core\CommandBus\Parser\CommandTypeParser;
+use PrestaShopBundle\CommandBus\MessengerCommandBusAdapter;
```
Comment (lartist) [inline src/Core/CommandBus/ExecutedCommandRegistry.php]: Sorry but we should not use PrestaShopBundle in the Core namespace :)

**Review by jolelievre — APPROVED** [review]
Thanks @M0rgan01 and @lartist

**Comment by jolelievre** [comment]
Running UI tests campaign https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/5508769177

**Review by matks — APPROVED** [review]

**Comment by jolelievre** [comment]
Thanks @M0rgan01 and @sallemiines
